# Question 1: Word2Vec on Health & Fitness Sentences

In this notebook we train a Word2Vec model on 20 health and fitness sentences using **Gensim**,
and then look at the words the model thinks are similar.

## Step 0: Install and import libraries

In [1]:
!pip install gensim


[notice] A new release of pip is available: 24.2 -> 26.2.1
[notice] To update, run: pip install --upgrade pip


In [2]:
import string
import nltk
import pandas as pd
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from gensim.models import Word2Vec

nltk.download('punkt')
nltk.download('punkt_tab')
nltk.download('stopwords')

[nltk_data] Downloading package punkt to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/sayanbhattacharya/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

## Step 1: Load the sentences into a Python list

In [3]:
sentences = [
    "A balanced diet helps maintain a healthy body weight",
    "Regular workout sessions improve overall muscle strength",
    "Drinking enough water daily supports good digestion",
    "A high protein diet is common among strength athletes",
    "Cardio workouts help improve heart and lung health",
    "Skipping meals can slow down your metabolism over time",
    "Stretching before a workout reduces the risk of injury",
    "A balanced diet should include fruits, vegetables and whole grains",
    "Strength training workouts build muscle and improve bone density",
    "Getting enough sleep is important for muscle recovery",
    "A sudden change in diet can cause digestive discomfort",
    "Yoga is a low-impact workout that improves flexibility",
    "Doctors recommend a balanced diet to prevent lifestyle diseases",
    "Interval training workouts burn calories quickly",
    "A high-sugar diet is linked to several health problems",
    "Consistent workout routines improve long-term fitness levels",
    "Nutritionists often design a diet plan based on individual goals",
    "Warm-up exercises prepare the body for an intense workout",
    "A vegetarian diet can still provide sufficient protein",
    "Rest days are an important part of any workout schedule"
]

print("Total number of sentences:", len(sentences))

Total number of sentences: 20


Save the corpus as a CSV file (needed for the submission).

In [4]:
df = pd.DataFrame({"text": sentences})
df.to_csv("word2vec_corpus.csv", index=False)
df.head()

,text
0,A balanced diet helps maintain a healthy body ...
1,Regular workout sessions improve overall muscl...
2,Drinking enough water daily supports good dige...
3,A high protein diet is common among strength a...
4,Cardio workouts help improve heart and lung he...


## Step 2: Convert to lowercase and remove punctuation

In [5]:
def clean_text(text):
    text = text.lower()
    for p in string.punctuation:
        text = text.replace(p, " ")   # replace punctuation with a space
    return text

cleaned_sentences = []
for s in sentences:
    cleaned_sentences.append(clean_text(s))

for s in cleaned_sentences[:5]:
    print(s)

a balanced diet helps maintain a healthy body weight
regular workout sessions improve overall muscle strength
drinking enough water daily supports good digestion
a high protein diet is common among strength athletes
cardio workouts help improve heart and lung health


## Step 3: Tokenize each sentence into words

In [6]:
tokenized_sentences = []
for s in cleaned_sentences:
    tokens = word_tokenize(s)
    tokenized_sentences.append(tokens)

for t in tokenized_sentences[:5]:
    print(t)

['a', 'balanced', 'diet', 'helps', 'maintain', 'a', 'healthy', 'body', 'weight']
['regular', 'workout', 'sessions', 'improve', 'overall', 'muscle', 'strength']
['drinking', 'enough', 'water', 'daily', 'supports', 'good', 'digestion']
['a', 'high', 'protein', 'diet', 'is', 'common', 'among', 'strength', 'athletes']
['cardio', 'workouts', 'help', 'improve', 'heart', 'and', 'lung', 'health']


## Step 4 (Optional): Remove stopwords

In [7]:
stop_words = set(stopwords.words('english'))

final_sentences = []
for tokens in tokenized_sentences:
    new_tokens = []
    for word in tokens:
        if word not in stop_words:
            new_tokens.append(word)
    final_sentences.append(new_tokens)

for t in final_sentences[:5]:
    print(t)

['balanced', 'diet', 'helps', 'maintain', 'healthy', 'body', 'weight']
['regular', 'workout', 'sessions', 'improve', 'overall', 'muscle', 'strength']
['drinking', 'enough', 'water', 'daily', 'supports', 'good', 'digestion']
['high', 'protein', 'diet', 'common', 'among', 'strength', 'athletes']
['cardio', 'workouts', 'help', 'improve', 'heart', 'lung', 'health']


In [8]:
# compare the number of unique words before and after stopword removal
words_before = set()
for tokens in tokenized_sentences:
    for w in tokens:
        words_before.add(w)

words_after = set()
for tokens in final_sentences:
    for w in tokens:
        words_after.add(w)

print("Unique words with stopwords   :", len(words_before))
print("Unique words without stopwords:", len(words_after))

Unique words with stopwords   : 122
Unique words without stopwords: 102


## Step 5: Train the Word2Vec model

The dataset is very small, so we use small values:
- `vector_size = 50` : each word becomes a list of 50 numbers
- `window = 3` : look at 3 words on the left and right
- `min_count = 1` : keep every word (even if it appears only once)
- `sg = 1` : use the skip-gram method (works better on small data)
- `epochs = 100` : go over the data many times because it is small
- `seed = 42` and `workers = 1` : so that we get the same result every time

In [9]:
model = Word2Vec(
    sentences=final_sentences,
    vector_size=50,
    window=3,
    min_count=1,
    sg=1,
    epochs=100,
    seed=42,
    workers=1
)

print("Model trained!")

Model trained!


## Step 6: Print the complete vocabulary

In [10]:
vocab = list(model.wv.index_to_key)

print("Total number of sentences:", len(sentences))
print("Vocabulary size:", len(vocab))
print()
print(vocab)

Total number of sentences: 20
Vocabulary size: 102

['diet', 'workout', 'improve', 'workouts', 'strength', 'muscle', 'balanced', 'important', 'training', 'health', 'protein', 'high', 'enough', 'body', 'schedule', 'part', 'days', 'rest', 'sufficient', 'provide', 'still', 'vegetarian', 'intense', 'prepare', 'exercises', 'warm', 'goals', 'individual', 'based', 'plan', 'design', 'often', 'nutritionists', 'levels', 'fitness', 'term', 'long', 'routines', 'consistent', 'problems', 'several', 'linked', 'sugar', 'quickly', 'calories', 'burn', 'interval', 'diseases', 'lifestyle', 'prevent', 'recommend', 'doctors', 'flexibility', 'improves', 'impact', 'low', 'yoga', 'discomfort', 'digestive', 'cause', 'change', 'sudden', 'recovery', 'sleep', 'getting', 'density', 'bone', 'build', 'grains', 'whole', 'vegetables', 'fruits', 'include', 'injury', 'risk', 'reduces', 'stretching', 'time', 'metabolism', 'slow', 'meals', 'skipping', 'lung', 'heart', 'help', 'cardio', 'athletes', 'among', 'common', 'diges

## Step 7: Word vector for "diet"

In [11]:
diet_vector = model.wv['diet']

print("Shape of the vector:", diet_vector.shape)
print()
print("First 10 values:")
print(diet_vector[:10])

Shape of the vector: (50,)

First 10 values:
[-0.02545768 -0.00867327  0.00423751 -0.00316094 -0.01257511  0.02292636
 -0.02652349  0.04279565 -0.03737486 -0.00703913]


In [12]:
print("Full vector for 'diet':")
print(diet_vector)

Full vector for 'diet':
[-0.02545768 -0.00867327  0.00423751 -0.00316094 -0.01257511  0.02292636
 -0.02652349  0.04279565 -0.03737486 -0.00703913 -0.01558221  0.04827917
 -0.00840396  0.03589156  0.03706356  0.01478708 -0.01585216 -0.04985502
  0.00947372 -0.00540012  0.00953812  0.00133883 -0.00629369  0.03325627
  0.04958679  0.00102725 -0.02828566 -0.00254329 -0.00377864 -0.0094717
 -0.00293793 -0.02196947 -0.0461235   0.01753233  0.00294092 -0.03195698
  0.01978445 -0.04920826 -0.00080506  0.03424937 -0.01213361  0.00041616
  0.04502997 -0.03477506 -0.00981525  0.01599505  0.01090025  0.03301707
  0.02598327 -0.00745531]


## Step 8: Top 5 words most similar to "diet"

In [13]:
similar_diet = model.wv.most_similar('diet', topn=5)

print("Top 5 words similar to 'diet':")
for word, score in similar_diet:
    print(word, ":", round(score, 4))

Top 5 words similar to 'diet':
improve : 0.7776
workout : 0.7495
based : 0.7402
enough : 0.7277
good : 0.7174


## Step 9: Top 5 words most similar to "workout"

In [14]:
similar_workout = model.wv.most_similar('workout', topn=5)

print("Top 5 words similar to 'workout':")
for word, score in similar_workout:
    print(word, ":", round(score, 4))

Top 5 words similar to 'workout':
enough : 0.835
improve : 0.8108
diet : 0.7495
workouts : 0.7464
muscle : 0.7144


Show both results in a small table using pandas.

In [15]:
table = pd.DataFrame({
    "diet - similar word": [w for w, s in similar_diet],
    "diet - score": [round(s, 4) for w, s in similar_diet],
    "workout - similar word": [w for w, s in similar_workout],
    "workout - score": [round(s, 4) for w, s in similar_workout]
})
table

,diet - similar word,diet - score,workout - similar word,workout - score
0,improve,0.7776,enough,0.8350
1,workout,0.7495,improve,0.8108
2,based,0.7402,diet,0.7495
3,enough,0.7277,workouts,0.7464
4,good,0.7174,muscle,0.7144


## Step 10: Interpretation

Some of the similar words make sense. For **workout**, the model found *workouts*, *muscle* and *improve*, which really
are related to exercise. For **diet**, words like *workout*, *improve* and *good* come up, which is partly sensible
(diet and workout are both about health), but words like *based* and *enough* are not really related to diet.
This happens because our dataset is very small (only 20 sentences and about 100 unique words), so most words appear
only once and the model does not see enough examples to learn good meanings. Also *diet* and *workout* appear in many
sentences, so they end up close to many common words. With a much bigger dataset the results would be cleaner.

## Questions for Interpretation

**1. Why is the `min_count` parameter needed, and what would happen if you increased it on this dataset?**

`min_count` tells the model to ignore words that appear fewer times than this number. In a big dataset this removes
rare words and spelling mistakes that the model cannot learn well anyway. In our small dataset most words appear
only once, so if we increase `min_count` the vocabulary becomes very small. As the check below shows, the vocabulary
drops from 102 words (min_count=1) to only 14 words (min_count=2) and 7 words (min_count=3), so we lose most of the words.

In [16]:
# small check: vocabulary size for different min_count values
for m in [1, 2, 3]:
    test_model = Word2Vec(sentences=final_sentences, vector_size=50, window=3,
                          min_count=m, sg=1, epochs=100, seed=42, workers=1)
    print("min_count =", m, "-> vocabulary size =", len(test_model.wv.index_to_key))

min_count = 1 -> vocabulary size = 102
min_count = 2 -> vocabulary size = 14
min_count = 3 -> vocabulary size = 7


**2. What does increasing the window size change about the relationships the model learns?**

The window is how many words on each side are treated as "context". A small window learns words that appear right
next to each other (more grammar-like / functional similarity). A larger window looks at more of the sentence, so the
model learns broader, topic-level relationships (for example, words from the same kind of sentence about diet or exercise).
In our short sentences a big window covers almost the whole sentence.

**3. Why might your similarity results differ slightly if you re-train the model?**

Word2Vec starts with random vectors and also uses random sampling (negative sampling) and, with several workers,
the order of training can change. So each training run can end up with slightly different vectors and scores.
With such a small dataset this effect is even bigger. Setting `seed` and `workers=1` helps to get the same result again.